### DISCUSSION TITLE
How to override get_bind() for AsyncSession?

### DISCUSSION THREAD

**Discussion by vytas7** [sqlalchemy/sqlalchemy#7383] [Usage Questions] [answered]
Hello,
I have been previously doing some vertical partitioning via overriding `Session.get_bind()` along the lines of [Custom Vertical Partitioning](https://docs.sqlalchemy.org/en/14/orm/persistence_techniques.html#custom-vertical-partitioning), and it works great.

However, I cannot seem to do the same for an `AsyncSession`, the overridden method is not invoked at all:
```python
import asyncio

from sqlalchemy import Column
from sqlalchemy import Integer
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import Session
from sqlalchemy.orm import declarative_base

Base = declarative_base()


class Object(Base):
    __tablename__ = 'objects'
    id = Column(Integer, primary_key=True)


class MySession(Session):
    def get_bind(self, mapper=None, clause=None):
        raise NotImplementedError('TODO: implement')
        # super().get_bind(mapper=mapper, clause=clause)


# session = MySession()
# session.add(Object())
# # Raises an instance of NotImplementedError as expected
# session.commit()


class MyAsyncSession(AsyncSession):
    def get_bind(self, mapper=None, clause=None):
        raise NotImplementedError('TODO: implement')
        # super().get_bind(mapper=mapper, clause=clause)


async def test_async_session():
    session = MyAsyncSession()
    session.add(Object())
    await session.commit()


asyncio.run(test_async_session())
```
... raises an instance of `sqlalchemy.exc.UnboundExecutionError` (_Could not locate a bind configured on mapper mapped class Object->objects, SQL expression or this Session_) after going via the vanilla `self.session.get_bind(bindkey, **kwargs)`.

As I understand, `AsyncSession` is some kind of a proxy, but how to control the underlying `get_bind()` in this scenario?

Should I instead override its `sync_session_class`?
This seems to work:
```python
session = MyAsyncSession(sync_session_class=MySession)
```
Is it the "correct" way of doing it?
The docs do mention that it is a proxied method, but maybe it should be stated that changing it on `AsyncSession` has no effect?

**Comment by zzzeek** [comment, marked as answer]
> 
> Should I instead override its `sync_session_class`? This seems to work:
> 
> ```python
> session = MyAsyncSession(sync_session_class=MySession)


yes

> ```
> 
> Is it the "correct" way of doing it? 

it is for now until we come up with something better, which would likely be an event hook to replace get_bind() totally


> The docs do mention that it is a proxied method, but maybe it should be stated that changing it on `AsyncSession` has no effect?

sure, we should override the method and put some docs to the effect "this method has no effect" and then maybe a section in the asyncio session docs overall regarding this.

**Reply by vytas7** [reply]
Thanks for shedding light on this!
(At first glance, I just happened to totally misunderstand `AsyncSession`'s proxying design. It makes much more sense now taking a second look at this.)

**Reply by CaselIT** [reply]
Btw you can also do it at class definition time:
```py
class MyAsyncSession(AsyncSession):
  sync_session_class=MySession
```

**Reply by vytas7** [reply]
A related thought though... what if one wanted to make `get_bind()` a coroutine function?
(I know it's a rather contrived scenario... maybe when trying to query health of different engines via some monitoring service :thinking: )

**Reply by CaselIT** [reply]
you don't need to. You can just provide an sync version and the db access will be transformed in async.
This would be the same as how the events work in asyncio. The even handler and the target are the sync versions, but the calls are not blocking from an even loop prospective

**Reply by zzzeek** [reply]
@CaselIT I guess we also need to illustrate the example showing the method still returning sync engines, i will put up a gerrit momentarily

**Reply by CaselIT** [reply]
Can't we deprecate it and create an event instead?

**Reply by vytas7** [reply]
Yes, that's understood.
But I thought of the case where one wanted to do async network I/O in `get_bind()` (probably one shouldn't), these calls would be blocking if they were async-unaware. Or are they offloaded to a threadpool?

**Reply by CaselIT** [reply]
they could use the more or less protected `util.concurrency.async_only` to await a co-routine. 
```py
async def go():
   async with aiohttp.ClientSession() as session:
      async with session.get('http://python.org') as response:
          return await response.text()
text = async_only(go())
```

**Reply by zzzeek** [reply]
this is the code example I'm doing, if we can confirm this works

```py
# using example from "Custom Vertical Partitioning"


import random

from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.ext.asyncio import create_async_engine
from sqlalchemy.orm import Session, sessionmaker

# construct async engines w/ async drivers
engines = {
    'leader':create_async_engine("sqlite+aiosqlite:///leader.db"),
    'other':create_async_engine("sqlite+aiosqlite:///other.db"),
    'follower1':create_async_engine("sqlite+aiosqlite:///follower1.db"),
    'follower2':create_async_engine("sqlite+aiosqlite:///follower2.db"),
}

class RoutingSession(Session):
    def get_bind(self, mapper=None, clause=None):
        # within get_bind(), return sync engines
        if mapper and issubclass(mapper.class_, MyOtherClass):
            return engines['other'].sync_engine
        elif self._flushing or isinstance(clause, (Update, Delete)):
            return engines['leader'].sync_engine
        else:
            return engines[
                random.choice(['follower1','follower2'])
            ].sync_engine

# apply to AsyncSession using sync_session_class
AsyncSessionMaker = sessionmaker(
    class_=AsyncSession,
    sync_session_class=RoutingSession
)
```

**Reply by CaselIT** [reply]
This is very similar to the example for ShardedSession here https://github.com/sqlalchemy/sqlalchemy/issues/7028

**Reply by zzzeek** [reply]
> Can't we deprecate it and create an event instead?

yes, but not today, and folks are looking for what to do for now, this is it :)  

an event here is a big deal because we have to maintain the old API and also get things like horizontal_shard to work and stuff like that.   might be a 2.1 thing if there is interest
